evmrpc/block.go

```diff
@@ -0,0 +1,109 @@
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/sei-protocol/sei-chain/x/evm/keeper"
+	"github.com/sei-protocol/sei-chain/x/evm/types"
+	rpcclient "github.com/tendermint/tendermint/rpc/client"
+	"github.com/tendermint/tendermint/rpc/coretypes"
+)
+
+type BlockAPI struct {
+	tmClient    rpcclient.Client
+	keeper      *keeper.Keeper
+	ctxProvider func() sdk.Context
+	txDecoder   sdk.TxDecoder
+}
+
+func NewBlockAPI(tmClient rpcclient.Client, k *keeper.Keeper, ctxProvider func() sdk.Context, txDecoder sdk.TxDecoder) *BlockAPI {
+	return &BlockAPI{tmClient: tmClient, keeper: k, ctxProvider: ctxProvider, txDecoder: txDecoder}
+}
+
+func (a *BlockAPI) GetBlockByHash(ctx context.Context, blockHash common.Hash, fullTx bool) (map[string]interface{}, error) {
+	block, err := a.tmClient.BlockByHash(ctx, blockHash[:])
+	if err != nil {
+		return nil, err
+	}
+	blockRes, err := a.tmClient.BlockResults(ctx, &block.Block.Height)
+	if err != nil {
+		return nil, err
+	}
+	return encodeTmBlock(a.ctxProvider(), block, blockRes, a.keeper, a.txDecoder, fullTx)
+}
+
+func encodeTmBlock(
+	ctx sdk.Context,
+	block *coretypes.ResultBlock,
+	blockRes *coretypes.ResultBlockResults,
+	k *keeper.Keeper,
+	txDecoder sdk.TxDecoder,
+	fullTx bool,
+) (map[string]interface{}, error) {
+	number := big.NewInt(block.Block.Height)
+	blockhash := common.HexToHash(string(block.BlockID.Hash))
+	lastHash := common.HexToHash(string(block.Block.LastBlockID.Hash))
+	appHash := common.HexToHash(string(block.Block.AppHash))
+	txHash := common.HexToHash(string(block.Block.DataHash))
+	resultHash := common.HexToHash(string(block.Block.LastResultsHash))
+	miner := common.HexToAddress(string(block.Block.ProposerAddress))
+	gasLimit, gasWanted := int64(0), int64(0)
+	transactions := []interface{}{}
+	for _, txRes := range blockRes.TxsResults {
+		gasLimit += txRes.GasWanted
+		gasWanted += txRes.GasUsed
+		decoded, err := txDecoder(txRes.Data)
+		if err != nil {
+			fmt.Println(err)
+			return nil, errors.New("failed to decode transaction")
+		}
+		if len(decoded.GetMsgs()) != 1 {
+			// EVM message must have exactly one message
+			continue
+		}
+		evmTx, ok := decoded.GetMsgs()[0].(*types.MsgEVMTransaction)
+		if !ok {
+			continue
+		}
+		ethtx, _ := evmTx.AsTransaction()
+		hash := ethtx.Hash()
+		if !fullTx {
+			transactions = append(transactions, hash)
+		} else {
+			receipt, err := k.GetReceipt(ctx, hash)
+			if err != nil {
+				continue
+			}
+			transactions = append(transactions, hydrateTransaction(ethtx, number, blockhash, receipt))
+		}
+	}
+	result := map[string]interface{}{
+		"number":           (*hexutil.Big)(number),
+		"hash":             blockhash,
+		"parentHash":       lastHash,
+		"nonce":            ethtypes.BlockNonce{}, // inapplicable to Sei
```

Comment:
as a meta point, i think we want to start documenting what is different b/w EVM on Sei vs EVM on ethereum? perhaps the PRD /RFC is a good place to start adding this, since these will inevitably come up later on and will be important for public facing documentation

Reply:
yeah good point. Added a section in https://docs.google.com/document/d/1WWagJLbUM_egr8gG_jhOHCADw2K11qgq83SWzPfZMzM/edit#heading=h.z0meha1ix02p